### ISSUE TITLE
注意力机制

### ISSUE THREAD

**Issue by ZY-LUCK** [guokan987/DGCN#1]
您好，我看了您的论文和代码，感觉您做的非常好，我学到了很多，但是代码中实现多头注意力机制这部分我看着不是很懂，想来请教您一下！希望您能看到，感谢！（附：邮箱1500647753@qq.com）

**Comment by ZY-LUCK** [comment]
您好，我看了您的论文和代码，感觉您做的非常好，我学到了很多，但是我没有看懂代码中怎么实现的多头注意力机制，想来请教您一下！希望您能看到，感谢！我的邮箱是1500647753@qq.com

**Comment by ZY-LUCK** [comment]
#SATT_3函数
f1 = self.conv1(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,3,1,4,2).contiguous()
f2 = self.conv2(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,1,3,4,2).contiguous()
logits = torch.einsum('bnclm,bcqlm->bnqlm',f1,f2)

**Comment by guokan987** [comment]
> #SATT_3函数
> f1 = self.conv1(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,3,1,4,2).contiguous()
> f2 = self.conv2(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,1,3,4,2).contiguous()
> logits = torch.einsum('bnclm,bcqlm->bnqlm',f1,f2)

你好，你的问题，我已知悉，这段代码就是多头注意力模型的实现方式，不知道你有什么疑问？

**Comment by guokan987** [comment]
> #SATT_3函数
> f1 = self.conv1(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,3,1,4,2).contiguous()
> f2 = self.conv2(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,1,3,4,2).contiguous()
> logits = torch.einsum('bnclm,bcqlm->bnqlm',f1,f2)

这里，我并没有用for循环的方式实现多头注意力，因为，不管是单头还是多头，其本质都是矩阵内积，所以你可以看见：如果我用单头，应该就没有“4”，直接在init中定义映射特征size为self.c_in//4；所以多头无非就是多了几次重复的动作，这里我做了4头注意力，所以我可以认为在特征映射阶段，init中定义映射特征size为self.c_in//4*4=self.c_in，再用这里的代码，分割出4个不同的特征，实现矩阵内积，这是一种实现的方式。

**Comment by guokan987** [comment]
> #SATT_3函数
> f1 = self.conv1(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,3,1,4,2).contiguous()
> f2 = self.conv2(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,1,3,4,2).contiguous()
> logits = torch.einsum('bnclm,bcqlm->bnqlm',f1,f2)

如果还有不明白，我把我的微信名字通过邮件发给你了，只能更高效的方式沟通了！😊

**Comment by shuqincao** [comment]
请问能加大佬微信吗？邮箱：1033898863@qq.com

**Comment by C3Shirt** [comment]
> > #SATT_3函数
> > f1 = self.conv1(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,3,1,4,2).contiguous()
> > f2 = self.conv2(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,1,3,4,2).contiguous()
> > logits = torch.einsum('bnclm,bcqlm->bnqlm',f1,f2)
> 
> 这里，我并没有用for循环的方式实现多头注意力，因为，不管是单头还是多头，其本质都是矩阵内积，所以你可以看见：如果我用单头，应该就没有“4”，直接在init中定义映射特征size为self.c_in//4；所以多头无非就是多了几次重复的动作，这里我做了4头注意力，所以我可以认为在特征映射阶段，init中定义映射特征size为self.c_in//4*4=self.c_in，再用这里的代码，分割出4个不同的特征，实现矩阵内积，这是一种实现的方式。

我用PEMS08数据运行了一下，这里的f1维度16(batchsize)*170(nodeNumber)*16*4*4  后面的三个维度16*4*4，其中16*4 是分出来4个特征，每个特征的长度是16 ，那还有一个4有什么实际意义吗？

**Comment by guokan987** [comment]
> > > #SATT_3函数
> > > f1 = self.conv1(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,3,1,4,2).contiguous()
> > > f2 = self.conv2(seq).view(shape[0],self.c_in//4,4,shape[2],shape[3]).permute(0,1,3,4,2).contiguous()
> > > logits = torch.einsum('bnclm,bcqlm->bnqlm',f1,f2)
> > 
> > 
> > 这里，我并没有用for循环的方式实现多头注意力，因为，不管是单头还是多头，其本质都是矩阵内积，所以你可以看见：如果我用单头，应该就没有“4”，直接在init中定义映射特征size为self.c_in//4；所以多头无非就是多了几次重复的动作，这里我做了4头注意力，所以我可以认为在特征映射阶段，init中定义映射特征size为self.c_in//4*4=self.c_in，再用这里的代码，分割出4个不同的特征，实现矩阵内积，这是一种实现的方式。
> 
> 我用PEMS08数据运行了一下，这里的f1维度16(batchsize)_170(nodeNumber)_16_4_4 后面的三个维度16_4_4，其中16*4 是分出来4个特征，每个特征的长度是16 ，那还有一个4有什么实际意义吗？

时间长度
